Add self-contained package with TeeDirectory implementation

magibney opened the pull request:
initially opening this against `fs/branch_9_3`; will switch the merge base branch to `fs/branch_9x` before merging.

The changes are nearly all circumscribed to the new package `org.apache.solr.storage`, with a handful of minor backports to support related functionality.

pvcnt committed:
SOLR-17094: Close objects contained inside an ObjectCache (#2110)

Co-authored-by: Vincent Primault <vprimault@salesforce.com>
(cherry picked from commit 1209027537bf929e5a0149eda92d76571ce936ac)

magibney committed:
close ObjectCache -- upstream will have this (or something like it) shortly

magibney committed:
support bypassing region cache (tmp commit)

magibney committed:
fix async cache stats when mapping function returns null

magibney committed:
Merge remote-tracking branch 'origin/fs/branch_9_3' into michaelgibney/close-object-cache

magibney committed:
update lucene hash (make ByteBufferGuard public, etc)

magibney committed:
ObjectCache entries should be closed if AutoCloseable, rather than Closeable

magibney committed:
DeleteCollectionCmd should default to removing index dir

magibney committed:
add TeeDirectory `storage` pacakage to solr fork

magibney committed:
tidy

magibney committed:
add license headers

magibney committed:
fix logging calls

magibney committed:
fix linting

magibney committed:
address errorprone checks

magibney committed:
run all tests with TeeDirectory by default

magibney committed:
in TeeDirectory, more robust/transparent exception handling

magibney committed:
adjust getCorePath() to getScopePath() -- handle snapshot/temp dirs

magibney committed:
TeeDirectory.listAll() must inspect both access and persistent

hiteshk25 commented on solr/core/src/java/org/apache/solr/storage/AccessDirectory.java:
```diff
@@ -0,0 +1,1687 @@
+      if (dir == null) {
+        return -1;
+      }
+      long offset = ((long) from[0]) << COMPRESSION_BLOCK_SHIFT;
```
Looks like from[0] has initialized with `0` value?  this would do nothing?

magibney replied:
`from[0]` maintains state across multiple calls to `LazyEntry.load()`. It's updated immediately below, each time it's passed to `LazyLoadInput` `in.load(long[])`

magibney committed:
ignore lazy temp files

magibney committed:
properly merge exposed pendingDeletes in TeeDirectory

hiteshk25 commented:
@magibney Still digesting. Couple of questions
1. Have we considered to extend `BufferedIndexInput` for `LazyLoadInput` ? I was looking `S3IndexInput` class, which extends the `BufferedIndexInput` class. it has override few methods `readInternal` and `seekInternal`, here it fills the buffer.
2. It seems general usage in lucene is `indeximput. IndexInput slice(String sliceDescription, long offset, long length)` api?

magibney committed:
minor bugfixes

magibney committed:
ByteBufferIndexInput sometimes requires extra buffers for some reason

so we pad the buffer array to avoid trouble

magibney committed:
some tests that manually create cores seems to have trouble with non-ephemeral dir impls

magibney committed:
we have to be more careful about access dir scoping in test context

magibney committed:
tidy

magibney commented:
> 1. Have we considered to extend `BufferedIndexInput` for `LazyLoadInput` ? I was looking `S3IndexInput` class, which extends the `BufferedIndexInput` class. it has override few methods `readInternal` and `seekInternal`, here it fills the buffer.

Yes, it looks like this could _almost_ work, or could be made to work (with some changes to upstream code), but I'd recommend against it for a few reasons:
1. The biggest problem is that file positioning in `BufferedIndexInput` is not block-aligned, which is really fundamental for `LazyLoadInput`
2. The other issue (which could be worked around) is that the `readInternal()` method assumes you're populating a buffer that's passed in, whereas in `LazyLoadInput`, we populate portions of much larger buffers and making them available, properly positioned, for use.
3. `LazyLoadInput` also implements basically all of the access optimizations in `ByteBufferIndexInput`, and `BufferedIndexInput` only implements a subset.

It would be reasonable to investigate making this change, but it involve a considerable amount of work, and I suspect the results would suffer in performance. Another reason I avoided going this route in the first place is that I was trying to minimize the entanglement with (and need to modify) upstream code that is unlikely to be changed upstream. `LazyLoadInput` definitely has some redundancy wrt `BufferedIndexInput`, in terms of the byte-level access implementations -- but then so does `ByteBufferIndexInput`. Many of these classes are pretty similar in the `readInt()` etc. implementations; what distinguishes them is how they interact with storage and buffers, and I think that the behavior of `LazyLoadInput` is different enough to where it makes sense to have it be a class that stands on its own. At least initially -- if we're still interested in exploring subclassing `BufferedIndexInput` maybe we could explore that in a subsequent iteration?


> 2. It seems general usage in lucene is `indeximput. IndexInput slice(String sliceDescription, long offset, long length)` api?

iiuc the question, basically yes. Lucene creates long-lived canonical IndexInputs over each file, then application code gets derivative copies of these via `slice()` and `clone()`, for use in single threads. All access to storage passes through these `IndexInput` objects.

hiteshk25 commented:
> 3\. ByteBufferIndexInput

Thanks @magibney . I think I understand your concern. do you feel readInt, readLong, readXXX methods are better than what `BufferedIndexInput` has? I feel that lazy code would run on VM-restart only. And if we can re-use `BufferedIndexInput` than it will simplify/tested our code. Based on this, would be good to follow up for atleast 1-day and see if we can extend `BufferedIndexInput` or other class

hiteshk25 commented:
>>then application code gets derivative copies of these via slice() and clone(), for use in single threads. All access to storage passes through these IndexInput objects.

In slice, do we we make sure we return the buffer with `length` only?

magibney commented:
> > 3. ByteBufferIndexInput
> 
> Thanks @magibney . I think I understand your concern. do you feel readInt, readLong, readXXX methods are better than what `BufferedIndexInput` has? I feel that lazy code would run on VM-restart only. And if we can re-use `BufferedIndexInput` than it will simplify/tested our code. Based on this, would be good to follow up for atleast 1-day and see if we can extend `BufferedIndexInput` or other class

It's no better, but on the other hand it's just not that complex really. For all that the `readInt()` etc. methods look complex, they're pretty rote/formulaic, so you'd be reducing the amount of fairly trivial code redundancy, at the expense of having our code more closely intertwined with upstream. In this case IMO the separation is actually a benefit -- and one could argue that subclassing could even be misleading, since we'd be subclassing `BufferedIndexInput` not so much because our classes would be a natural _subtype_, but simply to reduce the duplication of trivial code.

The extra methods that are overridden (following the template of `ByteBufferIndexInput`) do improve on `BufferedIndexInput`; and subclassing it, we would need to continue to override those anyway.

I could spend a day or so refactoring to do this and see how far I get; but even if we succeeded in the refactoring, such a broad refactoring could set us back wrt potentially invalidating the performance evaluation results that we have for the current version.

> In slice, do we we make sure we return the buffer with length only?

Yes, all the implementations of `slice` conform to the API spec.

hiteshk25 commented on solr/core/src/java/org/apache/solr/storage/CompressingDirectory.java:
```diff
@@ -0,0 +1,446 @@
+        LZ4.compressWithDictionary(compressBuffer, 0, 0, preBufferRemaining, out, ht);
+      }
+      int blockMapFooterSize = blockDeltas.transferTo(out);
+      if (wroteBlock) {
```
Can we have a separate methods for header/footer? 

1. What we are writing on those
2. Also, good to add byte for compression type (enum LZ4=1, gzip=2)

magibney replied:
They are really written together, so IMO it wouldn't make sense to separate them -- the header is the last thing written. The footer is written immediately for, by `int blockMapFooterSize = blockDeltas.transferTo(out)`, and then two things are written to the space pre-allocated for the header: the total logical length of the file, and the size of the serialized footer. The footer itself contains delta-encoded vint sizes of the individual compressed blocks, to allow mapping logical file blocks to their associated offset/length in the compressed file

magibney replied:
Sure, writing the compression type is a good idea -- really no reason not to!

hiteshk25 replied:
1. See if we can add blocksize as well. Also any other meta data
2. In general, function names are self documentation. Would be good to add documentation for that class, format, etc.

magibney replied:
a2ad9cf9f40ff673c7436aae2ba6c74bb2eee210 added compression type and block size (2 bytes), plus added 2 extra bytes for potential further extension without reindexing. The implementation barely changed (all static final fields, assuming the same type of compression), but the on-disk format records the metadata, the write code writes it, and the read code verifies it, so we can extend this to respect these values in a dynamic way later if/when need be.

hiteshk25 commented on solr/solrj/src/java/org/apache/solr/common/util/ObjectCache.java:
```diff
@@ -78,12 +84,26 @@ public <T> T computeIfAbsent(
   public void close() throws IOException {
-    isClosed = true;
-    map.clear();
+    if (isClosed.compareAndSet(false, true)) {
```
It's not clear to me changes in this file? is it separate issue?

magibney replied:
This was cherry-picked from upstream, and was initially the only change on this branch (before the TeeDirectory stuff moved in here). We depend on this to close node-level resources that are tracked in the ObjectCache (metrics, an ExecutorService, and if applicable the background eager loading task).

magibney replied:
https://github.com/cowpaths/fullstory-solr/pull/182/commits/654f882b601222238f411fa8e941fac62728e32e

hiteshk25 commented on solr/core/src/java/org/apache/solr/search/CaffeineCache.java:
```diff
@@ -211,6 +211,7 @@ private V computeAsync(K key, IOFunction<? super K, ? extends V> mappingFunction
       try {
         // Another thread is already working on this computation, wait for them to finish
         V value = result.join();
+        if (value == null) return null;
```
Separate issue?

magibney replied:
This is arguably a proper change in principle, but it's vestigial here in the sense that the need that caused this issue to become apparent is no longer relevant. I should remove this, and probably at some point pursue it upstream. Good catch.

magibney replied:
addressed with 51804e8d297b6e70e9152ac5db62a8f59aa074ee

hiteshk25 commented on solr/core/src/java/org/apache/solr/handler/RestoreCore.java:
```diff
@@ -99,44 +102,47 @@ public boolean doRestore() throws Exception {
-            if (cs == null) {
-              compareResult = new IndexFetcher.CompareResult();
-              compareResult.equal = false;
+      try (Lock lock = restoreIndexDir.obtainLock(IndexWriter.WRITE_LOCK_NAME)) {
```
What this change does?

magibney replied:
same as https://github.com/cowpaths/fullstory-solr/pull/182/files#r1508329085

hiteshk25 commented on solr/core/src/java/org/apache/solr/handler/IndexFetcher.java:
```diff
@@ -624,7 +628,15 @@ IndexFetchResult fetchLatestIndex(boolean forceReplication, boolean forceCoreRel
               .get(indexDirPath, DirContext.DEFAULT, solrCore.getSolrConfig().indexConfig.lockType);
 
-      try {
+      Lock lock = tmpIndexDir.obtainLock(IndexWriter.WRITE_LOCK_NAME);
```
not sure about this change as well?

magibney replied:
We only want to run TeeDirectory on the index dir, but there are other directories (that should just do normal file stuff) that are tracked through here. The reliable, if a bit ugly, way to detect this is to ensure that every directory that is being used for index purposes will have `obtainLock()` called on it. This always happens before the directory is read from or written to, except for in a few places where it was not strictly necessary because the directory reference is only held by local code (here and in `RestoreCore`).

hiteshk25 commented on solr/core/src/java/org/apache/solr/handler/component/SearchHandler.java:
```diff
@@ -389,6 +389,9 @@ protected boolean checkCircuitBreakers(
     return false;
   }
 
+  public static final ThreadLocal<Boolean> DISABLE_REGION_CACHE =
```
Not sure about this as well

magibney replied:
This is vestigial, it should be removed. Thanks for catching it!

magibney replied:
addressed with 8db36186b8ac96c74a44bc6627491f1346036b60

hiteshk25 commented on solr/core/src/java/org/apache/solr/cloud/api/collections/DeleteCollectionCmd.java:
```diff
@@ -116,6 +116,7 @@ public void call(ClusterState state, ZkNodeProps message, NamedList<Object> resu
       }
       ModifiableSolrParams params = new ModifiableSolrParams();
       params.set(CoreAdminParams.ACTION, CoreAdminParams.CoreAdminAction.UNLOAD.toString());
+      params.set(CoreAdminParams.DELETE_INDEX, message.getBool(CoreAdminParams.DELETE_INDEX, true));
```
what is for?

magibney replied:
This fixes a bug that is latent in upstream, but that our code depends on. Everywhere else that cores are deleted, the index dir is explicitly called to be deleted. This often has no practical effect, because the index dir is usually a subdir of the data dir, so it gets deleted implicitly. But because the index dir is tracked independently, it's clearly bad hygiene to not do this. It could probably just be hardcoded to `true` (as the params set immediately below), but here for now I opted to allow explicit override to preserve the previous behavior exactly.

magibney replied:
Without this change, the in-memory tracking of the directory is not closed out, which is a bit sloppy, but doesn't present a practical problem in most cases (because the in-memory tracking is about to get GC'd anyway). But we need to clean up extra associated paths (the access dir) so we rely on this hook (which, again, should really be there anyway in principle).

magibney committed:
more tolerant of missing persistent copy; verify persistent length asynchronously

magibney committed:
extra care wrt persitence for TeeDirectory.rename()

magibney committed:
avoid false positives for background length check on tmp files

magibney committed:
generic IOException should not trigger lazy file creation

otherwise it would in principle be possible to have both the
canonical file path _and_ an associated tmp file at the same
time

magibney committed:
update lucene hash post-lucene-merge

magibney committed:
Merge remote-tracking branch 'origin/fs/branch_9_3' into michaelgibney/close-object-cache

magibney committed:
expose accessDir path

magibney committed:
remove vestigial change in SearchHandler

magibney committed:
remove unrelated change in CaffeineCache; arguably a good change though

magibney committed:
write compression/block type info to header

magibney commented:
Thanks @hiteshk25! I think I've now addressed all your comments so far, if you care to review further?

magibney committed:
Directory.fileLength impl and invocations should align with spec for absent files

hiteshk25 reviewed (approved):
LGTM! (Thanks @magibney let's merge this)

magibney commented:
It's arbitrary whether I switch base branch to `fs/branch_9x` (as I'd initially planned to do) and backport to `fs/branch_9_3`, or squash-merge this to `fs/branch_9_3` and cherrypick/forward-port to `fs/branch_9x`. It will be cleaner and easier to do the latter, so that's what I'm going to do.
